AdvancedClusterStateTest failures

**changeClusterState_shouldFail_whenInitiatorDies_beforePrepare**
**changeClusterState_shouldFail_withoutBackup_whenInitiatorDies_afterPrepare**
**changeClusterState_shouldFail_withoutBackup_whenInitiatorDies_beforePrepare**

```
com.hazelcast.transaction.TransactionException: Transaction is timed-out!
    at com.hazelcast.transaction.impl.TransactionImpl.checkTimeout(TransactionImpl.java:287)
    at com.hazelcast.transaction.impl.TransactionImpl.prepare(TransactionImpl.java:211)
    at com.hazelcast.cluster.AdvancedClusterStateTest$DelegatingTransaction.prepare(AdvancedClusterStateTest.java:584)
    at com.hazelcast.cluster.AdvancedClusterStateTest$TransactionAnswer$1.prepare(AdvancedClusterStateTest.java:505)
    at com.hazelcast.cluster.impl.ClusterStateManager.changeClusterState(ClusterStateManager.java:210)
    at com.hazelcast.cluster.impl.ClusterServiceImpl.changeClusterState(ClusterServiceImpl.java:794)
    at com.hazelcast.cluster.AdvancedClusterStateTest.changeClusterState_shouldFail_whenInitiatorDies_beforePrepare(AdvancedClusterStateTest.java:160)
```

**invocationShouldComplete_whenMemberReJoins_inFrozenState**
**partitionTable_shouldBeFrozen_whenMemberLeaves_inFrozenState**

```
ava.lang.IllegalStateException: Still have pending migration/replication tasks, cannot lock cluster state! New state: FROZEN, current state: ACTIVE
    at com.hazelcast.cluster.impl.ClusterStateManager.lockClusterState(ClusterStateManager.java:124)
    at com.hazelcast.cluster.impl.operations.LockClusterStateOperation.run(LockClusterStateOperation.java:68)
```

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OracleJDK1.6/685/

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OpenJDK7/681/

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.x-OpenJDK6/684/

**clusterState_shouldBeFrozen_whenMemberReJoins_inFrozenState**
```
java.lang.AssertionError: Instance Member [127.0.0.1]:9114 this expected:<FROZEN> but was:<ACTIVE>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at com.hazelcast.cluster.impl.AdvancedClusterStateTest.assertClusterState(AdvancedClusterStateTest.java:541)
	at com.hazelcast.cluster.impl.AdvancedClusterStateTest.clusterState_shouldBeFrozen_whenMemberReJoins_inFrozenState(AdvancedClusterStateTest.java:445)
```
https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast/job/Hazelcast-3.x-IbmJDK1.8/com.hazelcast$hazelcast/115/testReport/junit/com.hazelcast.cluster.impl/AdvancedClusterStateTest/clusterState_shouldBeFrozen_whenMemberReJoins_inFrozenState/

**clusterState_shouldBeFrozen_whenMemberReJoins_inFrozenState**
```
java.lang.AssertionError: Instance Member [127.0.0.1]:7924 this expected:<FROZEN> but was:<ACTIVE>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at com.hazelcast.cluster.impl.AdvancedClusterStateTest.assertClusterState(AdvancedClusterStateTest.java:541)
	at com.hazelcast.cluster.impl.AdvancedClusterStateTest.clusterState_shouldBeFrozen_whenMemberReJoins_inFrozenState(AdvancedClusterStateTest.java:445)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.maintenance-IbmJDK1.8/com.hazelcast$hazelcast/76/testReport/junit/com.hazelcast.cluster.impl/AdvancedClusterStateTest/clusterState_shouldBeFrozen_whenMemberReJoins_inFrozenState/

**clusterState_shouldBeFrozen_whenMemberReJoins_inFrozenState**
```
java.lang.AssertionError: Instance Member [127.0.0.1]:5508 this expected:<FROZEN> but was:<ACTIVE>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:118)
	at com.hazelcast.cluster.impl.AdvancedClusterStateTest.assertClusterState(AdvancedClusterStateTest.java:541)
	at com.hazelcast.cluster.impl.AdvancedClusterStateTest.clusterState_shouldBeFrozen_whenMemberReJoins_inFrozenState(AdvancedClusterStateTest.java:445)
```

**invocationShouldComplete_whenMemberReJoins_inFrozenState**
```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertTrue(Assert.java:52)
	at com.hazelcast.cluster.impl.AdvancedClusterStateTest$11.run(AdvancedClusterStateTest.java:519)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:851)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:865)
	at com.hazelcast.cluster.impl.AdvancedClusterStateTest.invocationShouldComplete_whenMemberReJoins_inFrozenState(AdvancedClusterStateTest.java:516)
```

https://hazelcast-l337.ci.cloudbees.com/view/Official%20Builds/job/Hazelcast-3.maintenance-IbmJDK1.8/84/

--- a/hazelcast/src/main/java/com/hazelcast/transaction/impl/TransactionImpl.java
+++ b/hazelcast/src/main/java/com/hazelcast/transaction/impl/TransactionImpl.java
@@ -57,6 +57,7 @@ import static com.hazelcast.util.ExceptionUtil.rethrow;
 import static com.hazelcast.util.FutureUtil.ExceptionHandler;
 import static com.hazelcast.util.FutureUtil.RETHROW_TRANSACTION_EXCEPTION;
 import static com.hazelcast.util.FutureUtil.logAllExceptions;
+import static com.hazelcast.util.FutureUtil.waitUntilAllRespondedWithDeadline;
 import static com.hazelcast.util.FutureUtil.waitWithDeadline;
 import static com.hazelcast.util.UuidUtil.newUnsecureUuidString;
 import static java.lang.Boolean.TRUE;
@@ -228,7 +229,7 @@ public class TransactionImpl implements Transaction {
             createBackupLogs();
             state = PREPARING;
             List<Future> futures = transactionLog.prepare(nodeEngine);
-            waitWithDeadline(futures, timeoutMillis, MILLISECONDS, RETHROW_TRANSACTION_EXCEPTION);
+            waitUntilAllRespondedWithDeadline(futures, timeoutMillis, MILLISECONDS, RETHROW_TRANSACTION_EXCEPTION);
             state = PREPARED;
             replicateTxnLog();
         } catch (Throwable e) {

--- a/hazelcast/src/main/java/com/hazelcast/util/FutureUtil.java
+++ b/hazelcast/src/main/java/com/hazelcast/util/FutureUtil.java
@@ -26,6 +26,7 @@ import com.hazelcast.transaction.TransactionTimedOutException;
 
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
@@ -91,6 +92,23 @@ public final class FutureUtil {
         }
     };
 
+    private static final class CollectAllExceptionHandler implements ExceptionHandler {
+
+        private List<Throwable> throwables;
+
+        private CollectAllExceptionHandler(int count) {
+            this.throwables = Collections.synchronizedList(new ArrayList<Throwable>(count));
+        }
+
+        @Override
+        public void handleException(Throwable throwable) {
+            throwables.add(throwable);
+        }
+
+        public List<Throwable> getThrowables() {
+            return throwables;
+        }
+    }
 
     /**
      * Handler for transaction specific rethrown of exceptions.
@@ -243,6 +261,19 @@ public final class FutureUtil {
     @PrivateApi
     public static void waitWithDeadline(Collection<Future> futures, long timeout, TimeUnit timeUnit) {
         waitWithDeadline(futures, timeout, timeUnit, IGNORE_ALL_EXCEPT_LOG_MEMBER_LEFT);
+    }
+
+    @PrivateApi
+    public static void waitUntilAllRespondedWithDeadline(Collection<Future> futures, long timeout, TimeUnit timeUnit,
+                                                         ExceptionHandler exceptionHandler) {
+        CollectAllExceptionHandler collector = new CollectAllExceptionHandler(futures.size());
+        waitWithDeadline(futures, timeout, timeUnit, collector);
+        final List<Throwable> throwables = collector.getThrowables();
+        synchronized (throwables) {
+            for (Throwable t : throwables) {
+                exceptionHandler.handleException(t);
+            }
+        }
     }
 
     @PrivateApi

